src/PrestaShopBundle/EventListener/Admin/Context/ShopContextListener.php

```diff
@@ -183,4 +198,29 @@ private function redirectShopContext(RequestEvent $requestEvent): ?RedirectRespo
             ['setShopContext', 'conf']
         ));
     }
+
+    /**
+     * @throws ReflectionException
+     */
+    private function verifyAllShopAttribute(Request $request): ?ShopConstraint
+    {
+        try {
+            $routeInfo = $this->router->match($request->getPathInfo());
+            $controller = $routeInfo['_controller'];
+            [$className, $methodName] = explode('::', $controller);
+
+            $reflectionClass = new ReflectionClass($className);
+            $classAttributes = $reflectionClass->getAttributes(AllShopContext::class);
```

Comment:
Just a suggestion because I'm not sure it can work, but I believe Symfony has a `ClassMetadataFactoryInterface` service that handles the extraction of metadata from classes (among which the attributes) And since it's widely used in container compilation and DI there should be a cached version of this interface that may be used here and could be more performant than old school reflection that is rumored to be a bit heavy

I tested it simply by injecting the interface, it doesn't seem to have a heavy cache system but at least a local one in the system, so it should share the burden with other components relying on this

Oh actually it's only in dev mode, in prod mode the interface is implemented by `CacheClassMetadataFactory` so I guess it's worth the shot

Reply:
Thanks for sharing, but I don't think in our case it doesn't work for the methods attributes. I also looked, we use this on other listeners, and Symfony itself uses reflection to retrieve attributes https://github.com/symfony/http-kernel/blob/7.0/Event/ControllerEvent.php#L90